### PR TITLE
Fix races in ProducerSideServiceImplTest

### PR DESCRIPTION
**Pull request by dpallotti**
Two problems:
- `stream_` was destroyed before joining `read_thread_`, but it should be
  destroyed afterwards. Also wait before destroying `context_`, as it's also
  tied to the gRPC (http://b/207529277, http://b/207529528).
- `stream->Write` is not supposed to be called concurrently
  (http://b/185453746), but since we call
  `FakeProducer::SendBufferedCaptureEvents` and
  `FakeProducer::SendAllEventsSent` from different threads in these tests, add
  additional synchronization.

Bug: http://b/185453746
Bug: http://b/207529277
Bug: http://b/207529528

Test: Run `ProducerSideServiceImplTest.RedundantAllEventsSent` repeatedly before
the change: abort after a few hundred to thousand runs.
Run `ProducerSideServiceImplTest.RedundantAllEventsSent` 20000 times after the
change; also run all `ProducerSideServiceImplTest.*` 1000 times after the
change: no issues.

### CODE REVIEW

File: src/Service/ProducerSideServiceImplTest.cpp [hunk 1]
```diff
@@ -63,34 +74,56 @@ class FakeProducer {
...
     for (int32_t i = 0; i < num_to_send; ++i) {
       request.mutable_buffered_capture_events()->mutable_capture_events()->Add();
     }
+
+    is_writing_mutex_.LockWhen(absl::Condition(
+        +[](bool* is_writing) { return !*is_writing; }, &is_writing_));
+    is_writing_ = true;
     bool written = stream_->Write(request);
+    is_writing_ = false;
+    is_writing_mutex_.Unlock();
```
Comment (beckerhe) [inline src/Service/ProducerSideServiceImplTest.cpp:90]: Why isn't a simple mutex enough? Am I missing something?

Comment (dpallotti) [reply src/Service/ProducerSideServiceImplTest.cpp:90]: Oh boy, you're right...

Comment (dpallotti) [reply src/Service/ProducerSideServiceImplTest.cpp]: Done.

**Review by beckerhe — APPROVED** [review]
Awesome. Thank you! LGTM!

**Pull request merged**